# TASK 1
## Iris Flower Classification ML Project

## Importing Libraries

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Loading Dataset

In [4]:
df=sns.load_dataset('iris')
df

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

## Finding Missing Values

In [5]:
df.isnull().sum()

sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64

## Data Separation

In [6]:
x=df.iloc[:,0:4]
x

sepal_length  sepal_width  petal_length  petal_width
0             5.1          3.5           1.4          0.2
1             4.9          3.0           1.4          0.2
2             4.7          3.2           1.3          0.2
3             4.6          3.1           1.5          0.2
4             5.0          3.6           1.4          0.2
..            ...          ...           ...          ...
145           6.7          3.0           5.2          2.3
146           6.3          2.5           5.0          1.9
147           6.5          3.0           5.2          2.0
148           6.2          3.4           5.4          2.3
149           5.9          3.0           5.1          1.8

[150 rows x 4 columns]

In [7]:
y=df.iloc[:,4]
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: species, Length: 150, dtype: object

## Train Test Split

In [ ]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=22)
x_train

## Random Forest Classifier

In [9]:
from sklearn.ensemble import RandomForestClassifier
rfc=RandomForestClassifier(n_estimators=10) #by deafult trees=100
rfc

RandomForestClassifier(n_estimators=10)

In [18]:
rfc.fit(x_train,y_train)
y_pred=rfc.predict(x_test)
y_pred

array(['setosa', 'virginica', 'versicolor', 'virginica', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor', 'setosa',
       'virginica', 'versicolor', 'virginica', 'virginica', 'setosa',
       'virginica', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'setosa', 'virginica', 'setosa', 'versicolor',
       'virginica', 'setosa', 'virginica', 'virginica', 'virginica',
       'versicolor'], dtype=object)

#### Accuracy

In [12]:
from sklearn.metrics import accuracy_score
a=accuracy_score(y_test,y_pred)
a

0.9

#### Classification Report

In [13]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00         6
  versicolor       0.77      1.00      0.87        10
   virginica       1.00      0.79      0.88        14

    accuracy                           0.90        30
   macro avg       0.92      0.93      0.92        30
weighted avg       0.92      0.90      0.90        30



## KNN Classifier

In [14]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

In [20]:
knn=KNeighborsClassifier(n_neighbors=7)
knn
knn.fit(x_train,y_train)
y_pred=knn.predict(x_test)
y_pred

array(['setosa', 'virginica', 'versicolor', 'virginica', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor', 'setosa',
       'virginica', 'versicolor', 'virginica', 'virginica', 'setosa',
       'virginica', 'versicolor', 'versicolor', 'virginica', 'versicolor',
       'setosa', 'virginica', 'setosa', 'versicolor', 'virginica',
       'setosa', 'virginica', 'virginica', 'virginica', 'virginica'],
      dtype=object)

#### Accuracy

In [15]:
a=accuracy_score(y_test,y_pred)
print(a)

0.9666666666666667


#### Classification Report

In [16]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00         6
  versicolor       0.91      1.00      0.95        10
   virginica       1.00      0.93      0.96        14

    accuracy                           0.97        30
   macro avg       0.97      0.98      0.97        30
weighted avg       0.97      0.97      0.97        30



## Conclusion

### We have performed Iris Flower Classification using two classifiers namely Random Forest Classifier and KNN Classifier.
### Accuracy of Random Forest Classifier is 90%
### Accuracy of KNN Classifier is 96%
### Hence we conclude that KNN Classifier is better for classification of Iris Flower.